### PR TITLE
Nanotrasen Chefs have now officially gone to Chef College

### PR DESCRIPTION
**Pull request by FunnyMan3595**
## What Does This PR Do
(Cloned from #24422 with permission)

Roundstart chefs have a new spell that allows them to examine all items in 1 tile radius around them, and returns all the recipes they could make with that item. Its locked behind a 5 second cooldown due to performance concerns if someone spams it.

Roundstart chefs also can now examine a microwave/oven/other kitchen machinery to see if the items inside of it will make something, otherwise they won't get a message.

Also popcorn's gender is plural now, you're welcome.

## Why It's Good For The Game
Less dependency on the wiki is good.

## Images of changes
image
image
image
image

## Testing
See above

## Changelog
:cl:
add: Chefs have a new verb "Expert Chef Knowledge" that comes with their CQC, it allows them to examine food around them and see recipes they could make.
add: Chefs can also examine kitchen machinery (like microwaves) to see if the items in them will make anything.
fix: Popcorn now appears as "some popcorn" when examined.
/:cl:

### CODE REVIEW

File: code/datums/recipe.dm [hunk 1]
```diff
@@ -43,36 +43,84 @@
 	var/byproduct		// example: = /obj/item/kitchen/mould		// byproduct to return, such as a mould or trash
 
 /datum/recipe/proc/check_reagents(datum/reagents/avail_reagents) //1=precisely, 0=insufficiently, -1=superfluous
```
Comment (DGamerL) [inline code/datums/recipe.dm:45]:
```suggestion
/datum/recipe/proc/check_reagents(datum/reagents/avail_reagents)
```

File: code/datums/recipe.dm [hunk 2]
```diff
@@ -43,36 +43,84 @@
...
+
+		if(abs(provided_reagent_amount - reagents[required_reagent_id]) >= 0.5) // If amount is outside of the 0.5 unit tolerance
+			if(provided_reagent_amount > reagents[required_reagent_id]) // we have more than necessary
+				. = INGREDIENT_CHECK_SURPLUS
+			else
+				return INGREDIENT_CHECK_FAILURE // we don't have enough
+
+	if(length(reagents) < length(avail_reagents))
+		return INGREDIENT_CHECK_SURPLUS
 	return .
```
Comment (DGamerL) [inline code/datums/recipe.dm:80]:
```suggestion
	return 
```

File: code/datums/recipe.dm [hunk 3]
```diff
@@ -43,36 +43,84 @@
...
 				checklist -= type
 				found = 1
 				break
 		if(!found)
-			. = -1
+			. = INGREDIENT_CHECK_SURPLUS
 	if(checklist.len)
-		return 0
+		return INGREDIENT_CHECK_FAILURE
+	return .
```
Comment (DGamerL) [inline code/datums/recipe.dm:100]:
```suggestion
	return 
```

File: code/datums/recipe.dm [hunk 4]
```diff
@@ -43,36 +43,84 @@
...
+			return INGREDIENT_CHECK_SURPLUS
+
+		for(var/i in 1 to amount_given)
+			if(I in checklist)
+				checklist -= I
+			else if(I in items) // make sure the recipe requires this item before declaring it surplus
+				. = INGREDIENT_CHECK_SURPLUS
+	if(length(checklist)) // we didnt get everything
+		return INGREDIENT_CHECK_FAILURE
 	return .
```
Comment (DGamerL) [inline code/datums/recipe.dm:124]:
```suggestion
	return 
```

File: code/modules/food_and_drinks/kitchen_machinery/kitchen_machine.dm [hunk 5]
```diff
@@ -440,6 +440,20 @@
...
+		. += "<span class='notice'>You can't see inside the mixing bowl, you're not sure what it would do.</span>"
+		return
+
+	var/list/recipes = choose_recipes()
+	if(length(recipes) == 1 && recipes[1][2] != RECIPE_FAIL)
+		var/datum/recipe/recipe = recipes[1][2]
+		var/obj/item/result = recipe.result
+		. += "<span class='notice'>Your expert chef knowledge tells you that this would make \a [initial(result.name)].</span>"
+
+
```
Comment (DGamerL) [inline code/modules/food_and_drinks/kitchen_machinery/kitchen_machine.dm:456]:
```suggestion
```

File: code/game/jobs/job/support.dm [hunk 6]
```diff
@@ -243,6 +243,9 @@
 	var/datum/martial_art/cqc/under_siege/justacook = new
 	justacook.teach(H)
 	ADD_TRAIT(H.mind, TRAIT_TABLE_LEAP, ROUNDSTART_TRAIT)
+	ADD_TRAIT(H.mind, TRAIT_EXPERT_CHEF, ROUNDSTART_TRAIT)
+	if(H.mind)
+		H.mind.AddSpell(new /datum/spell/chef/expert_chef(null))
```
Comment (DGamerL) [inline code/game/jobs/job/support.dm:249]:
```suggestion
		H.mind.AddSpell(new /datum/spell/chef/expert_chef)
```
If you don't give a location, it'll automatically create it in nullspace

**Comment by Scribble-Sheep** [comment]
> I like this a lot mechanically, but I don't know how much I like the mention of a specific school.
Feels like that could mess with some character lore a bit, small of an issue as that sounds?

Reiterating this point from the original

**Comment by FunnyMan3595** [comment]
> > I like this a lot mechanically, but I don't know how much I like the mention of a specific school.
> > Feels like that could mess with some character lore a bit, small of an issue as that sounds?
> 
> Reiterating this point from the original

Updated the description, haven't bothered to update the PR image yet.

File: code/_globalvars/traits.dm [hunk 7]
```diff
@@ -81,7 +81,6 @@ GLOBAL_LIST_INIT(traits_by_type, list(
 		"TRAIT_NOSELFIGNITION_HEAD_ONLY" = TRAIT_NOSELFIGNITION_HEAD_ONLY,
 		"TRAIT_CONTORTED_BODY" = TRAIT_CONTORTED_BODY,
 		"TRAIT_DEFLECTS_PROJECTILES" = TRAIT_DEFLECTS_PROJECTILES,
-		"TRAIT_TABLE_LEAP" = TRAIT_TABLE_LEAP,
```
Comment (Contrabang) [inline code/_globalvars/traits.dm:84]: im not sure you meant to delete this

File: code/datums/spells/chef.dm [hunk 8]
```diff
@@ -0,0 +1,18 @@
+/datum/spell/chef/expert_chef/create_new_targeting()
+	return new /datum/spell_targeting/self
+
+/datum/spell/chef/expert_chef
+	name = "Expert Chef Knowledge"
+	desc = "Find things you can cook with the items in reach."
+	school = "chef"
+	panel = "Chef"
+	clothes_req = FALSE
+	base_cooldown = 5 SECONDS
```
Comment (Contrabang) [inline code/datums/spells/chef.dm:10]: if youre using a spell, I dont think you need the status effect anymore

File: code/_globalvars/traits.dm (hunk 7 above)
Comment (DGamerL) [reply code/_globalvars/traits.dm:84]: No, `TRAIT_TABLE_LEAP`is now a mind trait, this one was just never removed

**Comment by spacegeese** [comment]
would be really funny if using this ability with ice nearby told you you could use it in a deepfryer to create a firey explosion

File: code/datums/recipe.dm [hunk 9]
```diff
@@ -42,38 +42,86 @@
...
 			if(aval_r_amnt>reagents[r_r])
-				. = -1
+				. = INGREDIENT_CHECK_SURPLUS
 			else
-				return 0
+				return INGREDIENT_CHECK_FAILURE
 	if((reagents?(reagents.len):(0)) < avail_reagents.reagent_list.len)
-		return -1
+		return INGREDIENT_CHECK_SURPLUS
 	return .
```
Comment (DGamerL) [inline code/datums/recipe.dm:56]:
```suggestion
```

File: code/datums/recipe.dm [hunk 10]
```diff
@@ -42,38 +42,86 @@
...
+
+		if(abs(provided_reagent_amount - reagents[required_reagent_id]) >= 0.5) // If amount is outside of the 0.5 unit tolerance
+			if(provided_reagent_amount > reagents[required_reagent_id]) // we have more than necessary
+				. = INGREDIENT_CHECK_SURPLUS
+			else
+				return INGREDIENT_CHECK_FAILURE // we don't have enough
+
+	if(length(reagents) < length(avail_reagents))
+		return INGREDIENT_CHECK_SURPLUS
+	return
```
Comment (DGamerL) [inline code/datums/recipe.dm:80]:
```suggestion
```

File: code/datums/recipe.dm [hunk 11]
```diff
@@ -42,38 +42,86 @@
...
 				found = 1
 				break
 		if(!found)
-			. = -1
+			. = INGREDIENT_CHECK_SURPLUS
 	if(checklist.len)
-		return 0
-	return .
+		return INGREDIENT_CHECK_FAILURE
+	return
```
Comment (DGamerL) [inline code/datums/recipe.dm:100]:
```suggestion
```

File: code/datums/recipe.dm [hunk 12]
```diff
@@ -42,38 +42,86 @@
...
 		var/found = 0
 		for(var/type in checklist)
 			if(istype(O,type))
 				checklist -= type
 				found = 1
 				break
 		if(!found)
-			. = -1
+			. = INGREDIENT_CHECK_SURPLUS
 	if(checklist.len)
```
Comment (DGamerL) [inline code/datums/recipe.dm:98]:
```suggestion
	if(length(checklist))
```

File: code/datums/recipe.dm [hunk 13]
```diff
@@ -42,38 +42,86 @@
...
+			return INGREDIENT_CHECK_SURPLUS
+
+		for(var/i in 1 to amount_given)
+			if(I in checklist)
+				checklist -= I
+			else if(I in items) // make sure the recipe requires this item before declaring it surplus
+				. = INGREDIENT_CHECK_SURPLUS
+	if(length(checklist)) // we didnt get everything
+		return INGREDIENT_CHECK_FAILURE
+	return
```
Comment (DGamerL) [inline code/datums/recipe.dm:124]:
```suggestion
```

File: code/datums/recipe.dm [hunk 14]
```diff
@@ -42,38 +42,86 @@
...
 	for(var/r_r in reagents)
 		var/aval_r_amnt = avail_reagents.get_reagent_amount(r_r)
 		if(!(abs(aval_r_amnt - reagents[r_r])<0.5)) //if NOT equals
 			if(aval_r_amnt>reagents[r_r])
-				. = -1
+				. = INGREDIENT_CHECK_SURPLUS
 			else
-				return 0
+				return INGREDIENT_CHECK_FAILURE
 	if((reagents?(reagents.len):(0)) < avail_reagents.reagent_list.len)
```
Comment (DGamerL) [inline code/datums/recipe.dm:54]:
```suggestion
	if((reagents?(reagents.len):(0)) < length(avail_reagents.reagent_list))
```

File: code/__HELPERS/trait_helpers.dm [hunk 15]
```diff
@@ -232,6 +232,7 @@ Remember to update _globalvars/traits.dm if you're adding/removing/renaming trai
 #define TRAIT_TABLE_LEAP "table_leap" // Lets bartender and chef mount tables faster
 #define TRAIT_NEVER_MISSES_DISPOSALS "trait_never_misses_disposals" // For janitors landing disposal throws
 #define TRAIT_SLEIGHT_OF_HAND "sleight_of_hand"
+#define TRAIT_EXPERT_CHEF "expert_chef"
```
Comment (GDNgit) [inline code/__HELPERS/trait_helpers.dm:235]:
```suggestion
#define TRAIT_EXPERT_CHEF "expert_chef"
```
These are purely backend so it's the standard to keep their names functional. maybe something like TRAIT_KNOWS_COOKING_RECIPE

File: code/datums/recipe.dm [hunk 16]
```diff
@@ -42,38 +42,82 @@
...
 			if(aval_r_amnt>reagents[r_r])
-				. = -1
+				. = INGREDIENT_CHECK_SURPLUS
 			else
-				return 0
-	if((reagents?(length(reagents)):(0)) < length(avail_reagents.reagent_list))
-		return -1
-	return .
+				return INGREDIENT_CHECK_FAILURE
+	if((reagents?(reagents.len):(0)) < length(avail_reagents.reagent_list))
```
Comment (GDNgit) [inline code/datums/recipe.dm:54]:
```suggestion
	if((reagents ? length(reagents) : 0) < length(avail_reagents.reagent_list))
```

File: code/datums/spells/chef.dm [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+	human_req = TRUE
+	action_icon_state = "chef"
+	action_background_icon_state = "bg_default"
+	still_recharging_msg = "All this thinking makes your head hurt, wait a bit longer."
+
+/datum/spell/chef/expert_chef/cast(list/targets, mob/user = usr)
+	for(var/mob/living/carbon/human/H in targets)
```
Comment (GDNgit) [inline code/datums/spells/chef.dm:17]:
```suggestion
	var/mob/living/carbon/human/H = targets[1]
```
self targeting so we already know what's gonna be in here

File: code/modules/mob/living/carbon/human/human_mob.dm [hunk 18]
```diff
@@ -1962,3 +1962,68 @@ Eyes need to have significantly high darksight to shine unless the mob has the X
 /mob/living/carbon/human/proc/apply_offstation_roles(source)
 	SIGNAL_HANDLER
 	mind.offstation_role = TRUE
+
+/mob/living/carbon/human/proc/expert_chef_knowledge()
+	if(!HAS_TRAIT(mind, TRAIT_EXPERT_CHEF))
+		return
+
+	var/list/possible_cookware = orange(1, src)
```
Comment (GDNgit) [inline code/modules/mob/living/carbon/human/human_mob.dm:1970]:
```suggestion
	var/list/possible_cookware = view(1, src)
```
Orange and Oview ignore the center tile. Range ignores visibility of objects

File: code/modules/mob/living/carbon/human/human_mob.dm [hunk 19]
```diff
@@ -1962,3 +1962,68 @@ Eyes need to have significantly high darksight to shine unless the mob has the X
 /mob/living/carbon/human/proc/apply_offstation_roles(source)
 	SIGNAL_HANDLER
 	mind.offstation_role = TRUE
+
+/mob/living/carbon/human/proc/expert_chef_knowledge()
```
Comment (GDNgit) [inline code/modules/mob/living/carbon/human/human_mob.dm:1966]: Spooks me perfwise, can you check how long this function takes with like 20 valid objects?

**Review by GDNgit — CHANGES_REQUESTED** [review]

File: code/datums/spells/chef.dm (hunk 17 above)
Comment (Contrabang) [reply code/datums/spells/chef.dm:17]: counterpoint: xenomicrobes

Comment (GDNgit) [reply code/datums/spells/chef.dm:17]: Counterer point, requires a human to cast

**Review by GDNgit — APPROVED** [review]
lgtm

File: code/datums/spells/chef.dm [hunk 20]
```diff
@@ -0,0 +1,18 @@
+/datum/spell/chef/expert_chef/create_new_targeting()
+	return new /datum/spell_targeting/self
```
Comment (Contrabang) [inline code/datums/spells/chef.dm:2]: put this at the bottom of the file maybe? its strange that its at the top

**Review by Contrabang — DISMISSED** [review]

**Review by Burzah — APPROVED** [review]

File: code/modules/mob/living/carbon/human/human_mob.dm [hunk 21]
```diff
@@ -1967,3 +1967,68 @@ Eyes need to have significantly high darksight to shine unless the mob has the X
...
+
+	var/list/possible_cookware = view(1, src)
+
+	for(var/obj/item/storage/storage in possible_cookware)
+		possible_cookware += storage.contents
+		possible_cookware -= storage
+
+	shuffle_inplace(possible_cookware)
+
+
```
Comment (DGamerL) [inline code/modules/mob/living/carbon/human/human_mob.dm:1983]:
```suggestion
```

**Review by DGamerL — APPROVED** [review]

**Comment by BAGELMENSK** [comment]
This is actually a godsend

**Pull request merged**